runtime/compiler/trj9/runtime/ValueProfiler.cpp

```diff
@@ -188,13 +207,66 @@ TR_LinkedListProfilerInfo<TR_ByteInfo>::trace(TR::Compilation *comp)
    traceMsg(comp, "   Num: %d Total Frequency: %d\n", count, getTotalFrequency());
    }
 
+/**
+ * Lock access to the map, blocking if necessary.
+ */
+void
+TR_AbstractHashTableProfilerInfo::lock()
+   {
+   while (!tryLock())
+      {
+      // TODO: Does this need a yield?
+      }
+   }
+
+/**
+ * Try to lock access to the map.
+ *
+ * \param disableJITAccess This will trigger the short circuit test in JITed code.
+ */
+bool
+TR_AbstractHashTableProfilerInfo::tryLock(bool disableJITAccess)
+   {
+   // Set lock and, if desired, clear high bit in other index
+   MetaData locked = _metaData;
+   if (disableJITAccess)
+      locked.otherIndex &= ~HASHTABLE_LOCK_BIT;
+   locked.lock = 1;
+
+   // Current state must be unlocked
+   MetaData unlocked = _metaData;
+   unlocked.lock = 0;
+
+   return platformCompareSwap((uint32_t*) &_metaData, unlocked.rawData, locked.rawData);
+   }
+
+/**
+ * Unlock the map.
+ *
+ * \param enableJITAccess This will disable the short circuit test in JITed code.
+ */
+void
+TR_AbstractHashTableProfilerInfo::unlock(bool enableJITAccess)
+   {
+   // Clear lock and, if desired, set high bit in other index
+   MetaData unlocked = _metaData;
+   if (enableJITAccess)
+      unlocked.otherIndex |= HASHTABLE_LOCK_BIT;
+   unlocked.lock = 0;
+
+   // Overwrite what ever value is there
+   platformCompareSwap((uint32_t*) &_metaData, _metaData.rawData, unlocked.rawData);
+   }
+
 #if defined(TR_HOST_POWER)
 JIT_HELPER(_jitProfileValue);
 JIT_HELPER(_jitProfileLongValue);
 JIT_HELPER(_jitProfileBigDecimalValue);
 JIT_HELPER(_jitProfileStringValue);
 JIT_HELPER(_jitProfileAddress);
 JIT_HELPER(_jitProfileWarmCompilePICAddress);
+JIT_HELPER(_jProfileValue);
```

Comment:
should we call this jProfileIntValue? That way if we contribute it up to OMR where profiling of sub int values may be warranted the additions don't require refactoring?

Reply:
These have been corrected to use the `jProfileXXBitValue` naming.